\section{Methods}
For our method we assume that we have WGS data for both parental genomes and deep sequencing data of cell free DNA from maternal plasma. We model CNVs corresponding to a single parental haplotype duplication or deletion event. For each inheritance pattern (normal inheritance, maternal duplication, paternal duplication, maternal deletion, paternal deletion) we introduce a set of \emph{phased inheritance patterns} that enumerates all the possible configurations of fetal haplotypes corresponding to the respective inheritance pattern. E.g. for maternal duplication we have six phased inheritance patterns: $M_AM_AP_A, M_AM_BP_A, M_BM_BP_A, M_AM_AP_B, M_AM_BP_B, M_BM_BP_B$. This also enables us to reconstruct the origin of a CNV.

Our method models two types of signal from the data (i) change of the allele distributions at SNP loci, and (ii) change in number of fragments sequenced from a particular genomic region. Each of them is individual noisy but the two can be assumed independent for modelling purposes. Thus we combine them into one model. For this purpose we use a hidden Markov model, where we interpret the allele counts at SNP loci as emissions, and the coverage is used as a prior multiplied into transition probabilities. In the following we first describe each signal processing separately and then show how we combine them into a joint prediction of fetal CNVs.

\subsection{SNP Allele Distribution}\label{ss:allele_distrib}
For every SNP locus we observe a distribution of nucleotides in maternal plasma reads mapping to this particular position. Here we focus on calculating the probability of the observation w.r.t. a phased inheritance pattern. Formally, for observed 4-tuple $\(k_\A, k_\C, k_\G, k_\T\)$ of number of occurrences of each nucleotide, and phased inheritance pattern $\PP$ we model the conditional probability as Poisson distribution approximated by a Gaussian, i.e.
\begin{align}
Pr[\(k_\A, k_\C, k_\G, k_\T\) ~|~ \text{mat. haplotypes } M_A, M_B; \text{pat. haplotypes } P_A, P_B; \text{admixture } r; \PP] \sim \N\(\vmu, \vSigma \)
\end{align}
where
\begin{align*}
\vmu =&~ \(\mu_\A, \mu_\C, \mu_\G, \mu_\T\) \\
\vSigma =&~ \vmu \mathbf{I}_4
\end{align*}
To compute $\mu_x, x \in \{\A,\C,\G,\T\}$, we first adjust the mixture ratio $r$  based on phased pattern $\PP$ to reflect the expected number of fetal haplotypes $|H_\PP|$. 
\begin{align}
r' =&  \frac{ |H_\PP| \cdot r/2 }{|H_\PP| \cdot r/2 ~\cdot~ (1-r) }
\end{align}
Then for each nucleotide $x$ we sum probabilities of all the possible sources it might have been sequenced from, which includes maternal haplotypes and fetal haplotypes:
\begin{align}
p_x &= \sum_{i = A}^B [x \text{ equals } M_i] \cdot m_i (1-r')/2  \\
	&~~~ + \sum_{i = 1}^{|H_\PP|} [x \in H_\PP] \cdot r'/|H_\PP|  \nonumber
\end{align}
For reads putatively coming from indigenous maternal DNA, we correct for maternal CNVs by using the allele ratios $m_i$ as observed in maternal-only sequencing data. Additionally, in order to mitigate noise we add pseudocount $\alpha$ to these counts.
\begin{align}
m_i &= \frac{\alpha + \# \text{reads supporting }M_i\text{ in maternal seqencing}}{2\alpha + \sum_{j=A}^{B}\# \text{reads supporting }M_j\text{ in maternal seqencing}}
\end{align}
This way, we obtain the expected multinomial probability distribution over nucleotides aligned to this SNP locus in reads mapped to span this position. Thus to get the expected number of reads supporting particular variant at this SNP locus, we have to multiply $p_x$ by the number of reads mapped,
\begin{align}
\mu_x &= p_x \cdot \#\text{mapped reads}
\end{align}
As we describe later, we can use this probability distribution $\N\(\vmu, \vSigma \)$ that is conditional on phased pattern $\PP$ as an emission distribution in our HMM.
Note, that some observations have very similar probability under different phased patterns, e.g. some states of maternal deletion may yield distribution similar to paternal duplication. Incorporating the coverage signal into our main HMM helps to discriminate such states.

\subsection{CNVs and Depth of Coverage}\label{ss:coverage}
Variations in number of fragments sequenced per a region is a standard measure used for detection of mid to large sized CNVs\todo{citation}, and has lately been used for CNV detection from maternal plasma \cite{srinivasan2013, chen2013} as well. However the relatively low admixture of fetal DNA in the maternal plasma together with cell-free DNA sequencing biases considerably limit potential of methods relying on coverage signal from a single sample. Thus methods \cite{srinivasan2013, chen2013} require multiple datasets to establish a baseline for CNV calling.

In our method, we use the coverage information as a noisy predictor to complement the signal we obtain from SNP loci. For a reference plasma sequencing coverage we use plasma sample of the G1 trio of \cite{kitzman2012} dataset.

First, for each SNP $i$ we compute \emph{window ratio value} $\WRV_i$ for a window $W_i$ of size 1Kb centred to the $i$-th SNP. This measure is analogous to the \emph{bin ratio value} in \cite{srinivasan2013}, and we compute it as a ratio of number of fragments $N_{W_i}$ mapped to $W_i$ to sum of fragments mapped to 200 1Kb windows with $\G\C$ content closest to $W_i$
\begin{align}
\WRV_i &= \frac{ N_{W_i} }{ \mathlarger{\sum_{W \in neigh_{\G\C}^{200}(W_i)}} N_{W} }
\end{align}
Window ratio values are independent of $\G\C$ content and depth of sequencing, thus for a particular window they are directly comparable between different samples. We model the difference between $\WRV_i^S$ in the studied plasma sample and $\WRV_i^R$ in the reference plasma sample as a Gaussian noise with zero mean and empirically estimated variance $\sigma_{\text{noise}}$. 

Now we estimate the probability of the observed number of fragments $N_{W_i}$ in $W_i$ conditional on number of fetal haplotypes, which is either three for duplication, one for deletion, or two for normal inheritance. Therefore we compute two more $\WRV_i^R$s, each scaled to reflect one CNV type. For duplication, we would expect to see $(1+r/2)$ times more fragments while for deletion $(1-r/2)$ times less fragments, thus the scaled $\WRV_i^{R,DUP}$ and $\WRV_i^{R,DEL}$ are respectively estimated as
\begin{align}
\WRV_i^{R,DUP} = \frac{ N_{W_i^R} \cdot (1+r/2) }{ \mathlarger{\sum_{W \in neigh_{\G\C}^{200}(W_i^R)}} N_{W^R} }
\text{ ~~~~and~~~~ }
\WRV_i^{R,DEL} = \frac{ N_{W_i^R} \cdot (1-r/2) }{ \mathlarger{\sum_{W \in neigh_{\G\C}^{200}(W_i^R)}} N_{W^R} }
\end{align}

As mentioned earlier, our goal here is not to detect CNVs right away, but to rather compute a distribution over the number of haplotypes the fetus has inherited, which can later be used as a prior in our more complex model. In order to obtain these priors, we compute the posterior distribution in a hidden Markov model with three states -- duplication, deletion, and normal inheritance. For each state $s$ the emission probability of observed $\WRV_i^S$ is computed as $\N(\WRV_i^{R,s}-\WRV_i^S; \mu=0, \sigma_{\text{noise}})$. The HMM is depicted in Figure \ref{fig:hmm_coverage}.

\begin{figure}[h]
%\missingfigure{HMM for coverage}
\center\includegraphics[width = 0.25\textwidth]{figures/cov-HMM}
\caption{Hidden Markov model used to compute posterior distribution over number of haplotypes inherited by a fetus.}\label{fig:hmm_coverage}
\end{figure}

\subsection{Hidden Markov Model for CNV Inference}\label{ss:hmm}
To combine the signals form individual SNP positions, we use a hidden Markov model with states corresponding to modelled phased inheritance patterns, Figure \ref{fig:hmm_main}, totalling to 20 states. States representing normal inheritance are central to the model assuming that two CNVs cannot be immediately subsequent. Between states of the same inheritance pattern, we allow for transitions reflecting recombinations, but mainly errors in phasing.

For each state, an emission are the counts of individual alleles in reads mapped to that particular SNP position. The probability of the observed emission is the probability of such allele counts in the expected allele distribution conditional on phased pattern as describe above in \ref{ss:allele_distrib}.

\todo[inline]{explain the transitions and application of priors}

\begin{figure}[h]
%\missingfigure{The main HMM}
\centering
\subfigure[Overall HMM architecture]{
	\includegraphics[height=0.2\textheight]{figures/main-HMM}
	\quad
}
\hspace{20pt}
\subfigure[States for normal inheritance]{
	\begin{minipage}[b]{0.45\textwidth}
		\centering
		\includegraphics[height=0.2\textheight]{figures/NOR-HMM}
		\vspace{8pt}
	\end{minipage}	
}

\vspace{10pt}
\subfigure[States for maternal duplication]{
\begin{minipage}[b]{0.4\textwidth}
	\includegraphics[height=0.2\textheight]{figures/DUP-HMM}
	\end{minipage}	
}
\hspace{0pt}
\subfigure[States for maternal deletion]{
	\begin{minipage}[b]{0.4\textwidth}
		\centering
		\includegraphics[height=0.2\textheight]{figures/DEL-HMM}
	\end{minipage}	
}

\caption{Hidden Markov model used for CNV inference. We do not allow two CNVs to be adjacent, thus the switching always has to go through normal inheritance state.}\label{fig:hmm_main}
\end{figure}
