\section{Introduction}

Until recently, the prenatal analysis of a fetal genome required samples directly obtained from the fetus by invasive methods like amniocentesis. Such invasive methods are not effortable at high scale, because of increased health care expenses and also increased chance of miscarriage. Usually such medical intervention is done only in case of legitimate suspicion of a genetic disease, to confirm or reject a diagnosis. Therefore in the recent years, there has been interest in the development of methods that are minimally invasive. These methods analyze the DNA extracted from maternal plasma, which besides maternal DNA contains contains also admixture of fetal DNA. This admixture builds up from 5 to as much as 15 percent of the obtained material depending on the state of gravidity. In experiments conducted by \cite{kitzman2012} the estimated admixture in samples obtained at 18.5 weeks of gestation was 13\%, whereas \cite{bauer2006} report $<$10\%.

With the decreasing cost of DNA sequencing, these non-invasive methods are becoming more effortable and enable for preventive screening for genetic diseases, resulting in increase in quality of prenatal health care \cite{saunders2012}. Early diagnosis using non-invasive methods can shorten the time needed by differential diagnosis, which results in fewer empirical tests and faster progression in the treatment.

Until very recently, all the non-invasive prenatal genetic diagnostics aimed to test for previously identify diagnostic biomarkers of a specific disease. \cite{kitzman2012} is the first work that tackle the problem of non-invasive whole-genome sequencing of a fetus based on sequencing of both parental genomes and deep sequencing of maternal plasma (78-fold nonduplicate coverage). The main contribution of the authors to the whole-genome sequence reconstruction is in predicting of the set of alleles transmitted to the fetus from the parents at each SNP locus and novel mutations. 

To our best knowledge, so far there has not been published any non-invasive method on whole-genome analysis of CNVs that were de novo introduced in the fetal genome. The methods published so far aim for specific structural variances, like types of aneuploidy (abnormal number of chromosomes) \cite{chu2009}. This is our motivation for this project, in which we want to develop a method capable of calling also shorter CNVs than duplications (or deletions) of whole chromosomes.

The non-invasive methods (like \cite{kitzman2012, saunders2012, chu2009}) require deep sequencing of the maternal plasma, since the admixture is relatively low (\ntilde10\%). However even with 80-fold coverage, we expect so see only as few as 8 samples of fetal origin. Further, in case of normal inheritance, 4 samples of these 8 should come from a chromosome inherited by the fetus from the mother and the other 4 from an originally paternal chromosome. Authors in \cite{kitzman2012} exploit these ratios to predict fetal alleles at a given SNP locus. In case of maternal-only heterozygous sites (homozygous in father), they achieve only 64.4\% accuracy. To increase the accuracy (to 98.6\%) they require to have maternal genome phased to the individual haplotypes (i.e. for the heterozygous sites, in addition to the set of observed alleles, also to know which allele comes from which chromosome). In this project we are interested in structural, not sequential, variation. Thus our goal is not to identify the fetal alleles, but to rather identify from how many sources (copies of the sequence) the sequenced samples origin, and how many of them are inherited from maternal and paternal genome, respectively. 
