\section{Introduction}

Many genetic disorders, especially those associated with congenital malformations, are very difficult or impossible to treat. One of the most promising alternatives in such cases is prenatal screening of fetuses. 
Until recently, the prenatal analysis of a fetal genome required samples directly obtained from the fetus by invasive procedures like chorionic villus sampling or amniocentesis, where amniotic fluid is sampled from around the developing fetus.
Amniocentesis, however, has several important disadvantages. Foremost, it carries a non-trivial risk of miscarriage (estimated procedure-related fetal loss rate is 0.6\% to 1\% \citep{douglas2007amnio}), and hence is refused by a fraction of patients. Secondly, amniocentesis cannot be performed too early, as the risk of miscarriage rises significantly, and is typically indicated for the 15th week of pregnancy, outside of the time-frame for the safest abortion options ($<$12 weeks) and leaving only limited time for follow-up analysis. Finally, amniocentesis is a complex and expensive medical procedure (\$1,500--\$3,000). Consequently, amniocentesis is typically performed only to confirm or reject a diagnosis if a genetic disease is suspected, e.g. high likelihood of Down syndrome based on prenatal ultrasound. 

The last decade has seen the initial development of alternative, non-invasive methods for prenatal genetic testing. Prominent among these are methods that are based on analysis (arrays or sequencing) of  cell-free DNA (cfDNA) extracted from maternal blood plasma, which contains an admixture of fetal and maternal DNA. The fraction of fetal DNA in such an admixture varies depending on multiple factors, including maternal weight and size of the fetus, but typically builds up from \ntilde5-7\% early in the pregnancy to 10\% at week 10 \citep{wang2013} to as much as 50\% before delivery \citep{wang2013, fan2012}. In experiments conducted by \cite{kitzman2012} (and utilized in this paper) the estimated admixture in samples obtained at 8 weeks and 18.5 weeks of gestation was 7\% and 13\%, respectively. 

The decreasing cost of DNA sequencing has made it practical to directly sequence cfDNA extracted from maternal blood to identify likely genetic disorders present in the fetus.  Non-invasive methods are becoming more commonly used to directly identify aneuploidies (abnormal chromosome counts) and are also enabling preventive screening for heritable genetic diseases, resulting in increase in quality of prenatal health care \citep{saunders2012}.  While most non-invasive genetic diagnostics aim to test for a particular previously known biomarker,  \cite{kitzman2012} demonstrated the possibility of the reconstruction of the whole-genome of the fetus by combining whole-genome sequencing of both parental genomes with deep sequencing of cfDNA from maternal plasma (78x coverage). The key intuition in this method is the comparison of allelic ratios at individual SNP loci, as the inheritance of a particular paternal allele affects the percentage of reads with that allele at the particular position in the genome.  This method heavily relies on the availability of phased parental genotypes, as these allow for the inference of likely co-inherited SNPs, leading to an improvement in the signal-to-noise ratio. It consequently provides for high accuracy identification of inherited (98\% accuracy) but not \emph{de novo} single nucleotide variants (17 correct calls out of 44 true \textit{de novo} sites, with 3884 called positions).

The past year has seen the first few attempts at methods for genome-wide identification of sub-chromosomal \emph{de novo}  Copy Number Variation from cfDNA sequencing. While most of efforts have so far concentrated on whole-chromosome events (e.g. \cite{chu2009}), two manuscripts address the problems of detecting sub-chromosomal CNVs \citep{chen2013, srinivasan2013}. While the exact methods used in both of these approaches differ, both rely on depth of coverage:  they map the reads to the genome, divide the genome into bins, and identify the CNVs by comparing the number of reads mapped to each bin. The key idea in these methods is that deletions/duplications will result in more/fewer fetal reads within a window, and this difference can be identified using statistical methods. \cite{srinivasan2013} use depth-of-coverage computed in 1Mb windows across the genome to identify CNVs that are typically $>$1MB, though they do report discovery of a 300kb CNV. 9 of the 22 discovered CNVs in 11 patients were concordant with karyotyping results, with most discrepancies being short ($<$1Mb) CNVs. Importantly, they use extremely short (25bp) reads, allowing for larger number of fragments at equal coverage depth. \cite{chen2013} use even larger 10Mb windows, again considering only the number of fragments mapped  and are able to successfully identify variants 9-29Mb with only one false positive among 6 true positives in 1311 patients.

In this manuscript we introduce a novel model for non-invasive prenatal identification of de novo CNVs with largely increased sensitivity compared to methods published so far. Our method combines three types of information within a unified probabilistic model. First, our method takes advantage of the imbalance of allelic ratios at SNP positions that are introduced by various types of paternally and maternally inherited CNVs. Secondly, following the work of \cite{kitzman2012}, we use parental genotypes to phase nearby SNPs, modelling their co-inheritance (or recombination) and thus improving the signal-to-noise ratio. Finally, we observed that allelic ratios poorly differentiate between certain types of CNVs: for example, as further described below,  a duplication of a paternally inherited allele results in extremely similar allelic ratios to deletion of a maternally inherited one. We thus combine the allelic ratios with the depth-of-coverage signal to better differentiate between such cases. Our simulation results, based on \emph{in silico} introduction of novel CNVs into plasma samples with 13\% fetal DNA concentration, demonstrate a sensitivity of 90\% for CNVs $>$400 kilobases (with 13 calls in an unaffected genome),  and 40\% for 50-400kb CNVs (with 108 calls in an unaffected genome). 


