\section{Introduction}

The past 6 years have seen the development of methodologies to identify genomic variation within a fetus through the sequencing of maternal blood plasma. These novel non-invasive methods are based on the observation that maternal plasma contains a fraction of DNA (typically 5-15\%) originating from the fetus \citep{lo1997presence}. The key advantage of non-invasive methods is that they pose no associated risk to the pregnancy. In contrast, invasive methods like chorionic villus sampling or amniocentesis (sampling of amniotic fluid from around the developing fetus) have estimated procedure-related fetal loss rate of 0.6\% to 1\%, \cite{douglas2007amnio}. Non-invasive methods have already been used, mainly for the detection of whole-chromosomal events like trisomy of chromosome 21 (Down syndrome) \citep{chiu2008noninvasive, fan2008noninvasive}, and to a more limited extent for smaller (typically several megabases long) Copy Number Variants (CNVs) \citep{chen2013, srinivasan2013, rampasek2014fcnv}. Methods utilizing genome-wide detection of sub-chromosomal CNVs are highly desired, to enable prenatal screening for diseases like DiGeorge syndrome (\ntilde3Mb deletion), Prader-Willi syndrome (\ntilde4Mb deletion), and other, associated with a mid to large sized CNV. In addition to CNV detection, a successful proof-of-concept for non-invasive genome-wide fetal genotyping has been published \citep{kitzman2012}. This is the only work (to our best knowledge), in the field of maternal plasma analysis for fetal variation detection, with publicly available data set.

In this manuscript we introduce a new method for non-invasive fetal CNV detection given parental genomes and deep sequencing of maternal plasma cell free DNA (cfDNA). We build upon Hidden Markov Model based method by \cite{rampasek2014fcnv}, seeking to address several of its modelling drawbacks. Foremost, we employ a discriminative model, Conditional Random Field, that helps to focus on the classification goal and enables us to better incorporate multiple dependant data features. For training and testing of our model, we utilize the \textit{in silico} simulated CNV data set by \cite{rampasek2014fcnv} based on sequencing samples of \cite{kitzman2012}.

In the following, we give a brief description of the methods published so far, provide motivation for employing CRF rather than HMM for the CNV detection task, introduce our model in detail, and present the obtained results.

\subsection{Existing methods}
There have been three methods published so far that try to address the aforementioned problem of whole-genome detection of sub-chromosomal copy number variants in fetal genome from maternal plasma cfDNA sequencing. These can be grouped into two following categories.

\subsubsection{Methods utilizing depth-of-coverage signal}
Two methods published last year \citep{chen2013, srinivasan2013} utilize only low-coverage high-throughput sequencing data of cfDNA from maternal plasma. Such sequencing is cheap (in order of few hundred USD per patient) what makes them well suited for practice in this regard.

While the exact methods used in both of these approaches differ, the essential idea remains the same. They map the reads to the reference genome, divide the genome into bins, and identify the CNVs by comparing the number of reads mapped to each bin with values obtained in control samples. The key idea in these methods is that deletions/duplications will result in more/fewer fetal reads within a window, and this difference can be identified using statistical methods.

In \cite{srinivasan2013} authors use depth-of-coverage computed in 1Mb windows across the genome to identify CNVs that are typically at least 1Mb long, though they do report discovery of a 300kb CNV. 9 of the 22 discovered CNVs in 11 patients were concordant with karyotyping results, with most discrepancies being short ($<$1Mb) CNVs.

In \cite{chen2013} authors use the same WG sequencing samples as are currently used for trisomy detection. Average DOC of such samples is less than one. Thus they use even larger 10Mb windows, again considering only the number of fragments mapped and are able to successfully identify variants 9-29Mb with only one false positive among 6 true positives in 1311 patients.

The limiting factor of these methods is the relatively large minimal CNV length  necessary for successful detection. However \cite{srinivasan2013} claims resolution down to 1Mb or even less, the paper provide only very limited justification, as only 3 from 16 calls of size $<$2Mb were true positives.

\subsubsection{Methods incorporating multiple signals}
In a recent work \citep{rampasek2014fcnv} (currently in publication process) we introduced a probabilistic method for whole-genome non-invasive analysis of \textit{de novo} CNVs in fetal genome based on maternal plasma sequencing with largely increased sensitivity compared to methods published before. The method combines three types of information within a unified Hidden Markov Model: the imbalance of allelic ratios at SNP positions; the use of parental genotypes to phase nearby SNPs; and depth of coverage to better differentiate between various types of CNVs and to improve precision. Simulation results, based on \emph{in silico} introduction of novel CNVs into plasma samples with 13\% fetal DNA concentration, demonstrate a sensitivity of 90\% for CNVs $>$400 kilobases (with 13 calls in an unaffected genome), and 40\% for 50-400kb CNVs (with 108 calls in an unaffected genome).