\pdfoutput=1
\documentclass{bioinfo}
\copyrightyear{2014}
\pubyear{2014}

\usepackage{natbib}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{amsmath}
\usepackage{subfigure}
\usepackage{listings}
\usepackage{url}
\usepackage{relsize}
\usepackage{multirow}
\usepackage{bm}
% nice tilde
\newcommand{\ntilde}{\raise.17ex\hbox{$\scriptstyle\mathtt{\sim}$}}

\newcommand{\refe}{R}
\newcommand{\alte}{A}
\newcommand{\pos}{t}
\newcommand{\ar}{x}
\newcommand{\pip}{y}
\newcommand{\pips}{\mathbf{\pip}}
\newcommand{\ars}{\mathbf{\ar}}
\newcommand{\indicator}[1]{\bm{[}#1\bm{]}}

\DeclareMathOperator*{\argmax}{arg\,max}
\DeclareMathOperator*{\copycount}{cc}

\def\inst#1{${}^{#1}$}

\begin{document}

\title[Application of conditional random fields for non-invasive fetal CNV detection from maternal plasma]{Application of conditional random fields for non-invasive fetal CNV detection from maternal plasma}
\author[Maddison and Rampasek]{
	Chris J. Maddison\inst{1}, and
	Ladislav Rampasek\inst{1}
	}

\address{
\inst{1}
    Department of Computer Science, University of Toronto, Toronto M5S 2E4, Canada
}

\history{}
\editor{}


\maketitle

\begin{abstract}
\section{Motivation:}
Non-invasive methods facilitating genome-wide detection of sub-chromosomal CNVs in fetal genome are highly desired, to enable prenatal screening for congenital genetic diseases. Non-invasive prenatal methods pose no side-effects or method-related health risks for neither the mother or its fetus. Such methods rely on sequencing of cell-free DNA contained in maternal plasma, which is known to contain admixture of fetal-derived DNA during the pregnancy. Although detection of whole-chromosomal copy variations has been extensively studied in past years, there has been little work published on generic genome-wide sub-chromosomal CNVs detection. Here, we introduce a new model utilizing Conditional Random Fields to leverage multitude of data-derived signals to address this task.
\section{Results:}
We have developed a CRF utilizing the following features: parental haplotype phasing, number of reads supporting reference or alternative alleles at 1000gp SNP loci modelled by beta-binomial distribution, depth-of-coverage around these SNP loci, and distance between the adjacent SNP loci. We approached training of the CRF with two different methods, albeit with limited success.
\section{Availability:}
http://github.com/rampasek/FetalCNV

\section{Contact:}
\path|{cmaddis, rampasek}@cs.toronto.edu| 
\end{abstract}


\input intro
\input methods
\input results
\input discussion

%\subsection*{Acknowledgements}

\bibliographystyle{natbib}
\bibliography{main}

\end{document}
